### DISCUSSION TITLE
struggling to bind brightness keys to brillo commands using i3/sxhkd

### DISCUSSION THREAD

**Discussion by ryan-bachir** [i3/i3#6194] [Q&A]
ive got a laptop and i want to bind the brightness keys to increase and decrease the laptop's brightness but im running into trouble. im using brillo to change the brightness, which i have confirmed as working through the terminal, however it does not seem to work when bound to the brightness keys.

at first i tried using bindsym to bind the brightness keys however that didnt do anything. i then tried using sxhkd which also did not seem to work when i ran sxhkd from my i3 config.

HOWEVER when i ran sxhkd from the terminal after removing sxhkd from my i3 config, the brightness keys did work. and when i added sxhkd back to my i3 config and tried running sxhkd after that, it told me that it ran into an error because the keys were already bound.

so essentially, sxhkd is working when running it from the terminal and is binding the brightness keys correctly, however when i try to run sxhkd from i3 using `exec sxhkd &` or `exec sxhkd` or `exec sxhkd -c ~/path/to/sxhkdconfig` or whatever other variation, it does not seem to work (despite it detecting the keys as being bound when running after the fact)

any suggestions? im really not sure whats not working here

**Comment by okraits** [comment]
It's difficult to give advice without looking at your actual config. But I'll try to give some hints:

- don't map the same keys in both i3 and sxhkd
- kill sxhkd when mapping keys in i3 to avoid conflicts between the two

As a reference, I'm using the following mappings:

```
bindsym XF86MonBrightnessDown exec xbacklight -dec 10 && pkill -o -RTMIN+2 i3blocks
bindsym XF86MonBrightnessUp exec xbacklight -inc 10 && pkill -o -RTMIN+2 i3blocks
```

**Reply by ryan-bachir** [reply]
i have been making sure that both i3 and sxhkd arent mapping the brightness keys simultaneously, and i have been rebooting to ensure sxhkd is fully killed when testing with bindsym, but the problem still remains. when in i3 i have been using `bindsym XF86MonBrightnessUp exec brillo -A 10` for example, but that doesnt seem to work. when i remove that and try running sxhkd from my config, it still doesnt work. however, when i remove both the bindsym and sxhkd from my i3 config and run sxhkd directly from terminal, it does work properly. so i think it has to do with the i3 config itself but i dont know why it wouldnt work when run from the i3 config but does work when run outside of it

**Reply by okraits** [reply]
What happens if you bind a command known to work to `XF86MonBrightnessUp`, for example to execute your terminal emulator? Does that work?

**Reply by ryan-bachir** [reply]
so binding other commands do seem to work. i also did some more testing and i have figured out the problem and now just need to figure out the solution. i will outline everything ive done just for clarity.

i made sure all instances of sxhkd were off and that `XF86MonBrightnessUp` was not being bound by i3, sxhkd, or anything else. i then readded it to my sxhkd with the `alacritty` command and ran sxhkd from i3 by putting `exec --no-startup-id sxhkd &` in my i3 config. this did work, so i then replaced it with `brillo -A 10` and that also seemed to work properly. now the key thing is that i have a bash script with dunst which gives me a notification when i change the brightness, which is as follows:
```
#!/bin/bash

function send_notification() {
    brightness=$(printf "%.0f\n" $(brillo -G))
    dunstify -a "changebrightness" -u low -r 9991 -h int:value:"brightness" -i "brightness-$1" "Brightness: $brightness%"
}

case $1 in 
up)
    brillo -A 5
    send_notification $1
    ;;
down)
    brillo -U 5
    send_notification $1
    ;;
esac
```

this script works fine when i run it in terminal by using `changebrightness up` for example, however when i try to add this to my sxhkd config it does not do anything. the same goes for when i try using bindsym instead of sxhkd. i did manage to get it to work by doing `path/to/command/changebrightness up` instead of just `changebrightness up` but i am curious now as to why the bash script doesnt work when run from i3 config/sxhkd but does work in terminal/when i run it from the path rather than as a command.

**Reply by okraits** [reply]
One hint: you don't need the `&` at the end of `exec --no-startup-id sxhkd &` - i3 forks when running external commands.

If you need to use the full path to your script, that means that this path is not in the `PATH` of the shell used by i3/sxhkd.

**Reply by ryan-bachir** [reply]
thanks for the hint. i also just decided not to both with adding my scripts path to the shell path and just run them as i did above, but might look into adding my scripts path at a later point. thanks again!
